DateHistogram: Fix `extended_bounds` with `offset`

cbuescher opened the pull request:
This fixes a bug in 'date_histogram' when using 'extended_bounds'
together with some 'offset'. Offsets should be applied after rounding the
extended bounds and also be applied when adding empty buckets during the reduce
phase in InternalDateHistogram.

Closes #23776

cbuescher committed:
DateHistogram: Fix 'extended_bounds' with 'offset'

This fixes a bug in 'date_histogram' when using 'extended_bounds'
together with some 'offset'. Offsets should be applied after rounding the
extended bounds and also be applied when adding empty buckets during the reduce
phase in InternalDateHistogram.

Closes #23776

jpountz commented on core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/ExtendedBounds.java:
```diff
@@ -166,8 +166,9 @@ ExtendedBounds parseAndValidate(String aggName, SearchContext context, DocValueF
-        return new ExtendedBounds(min != null ? rounding.round(min) : null, max != null ? rounding.round(max) : null);
+    ExtendedBounds round(Rounding rounding, long offset) {
+        return new ExtendedBounds(min != null ? rounding.round(min) + offset : null,
+                max != null ? rounding.round(max) + offset : null);
```
should it be `rounding.round(X - offset) + offset` instead?

cbuescher replied:
I don't think so. The min/max values should initially be the ones defined by the user in the query (e.g. "2016-01-01T04:00:00Z"). I we subtract the offset before rounding and e.g. do 1day intervals, we would get to 2015-12-31 in the example above, which I think is not the intention of the user. Does this make sense or am I missing something?

jpountz reviewed (commented):
I left a question.

cbuescher commented:
@jpountz I responded to your question, not sure if this makes sense. I could do the additional offset calculation in InternalDateHistogram#addEmptyBuckets() but this way its more centralized.

jpountz commented:
I think part of the problem here is that the extended bounds are not well defined, eg. what is the expected behaviour if the provided bounds are not rounded values, should we round up or down? Maybe we should open a `discuss` issue about this.

cbuescher commented:
@jpountz I opened #24064 to discuss this, but I think the current behaviour of always rounding down `extended_bounds` makes a lot of sense to me.

cbuescher committed:
Don't add offset in ExtendedBounds#round()

cbuescher commented:
I rebased and changed this PR a bit so the offset isn't added in `ExtendedBounds#round()` but in `InternalDatehistogram#addEmptyBuckets`. I changed this because otherwise `ExtendedBounds#round()` wouldn't be an idempotent method, so mistakenly calling it several times with a set offset would have the potential to move the bounds further than intended. I think it is safer to keep the min/max bounds to be the rounded values and only fix the bug in the missing offset calculation in `addEmptyBuckets`
@jpountz could you have another look if this is better now?

kujon commented:
Could you please clarify which release was it included in in the end?

cbuescher commented:
@kujon the attached labels indicate which versions this was merged to. So 5.5.0 means it is included in 5.5.x and all newer 5 version, and 6.0.0-alpha means its on all current 6.x releases. Hope this helps.
